Implemented Tanimoto distance and similarity metrics

StaniszewskiA opened the pull request:
## Changes

*Short description of changes*

## Checklist before requesting a review
- [x] Docstrings added/updated in public functions and classes
- [x] Tests added, reasonable test coverage (at least ~90%, `make test-coverage`)
- [x] Sphinx docs added/updated and render properly (`make docs` and see `docs/_build/index.html`)

StaniszewskiA commented:
I left **from sklearn.utils._param_validation import validate_params** inside **similarities/tanimoto.py** since I am unsure whether to use it or not. Let me know if I should incorporate this type of validation in the code, and I will make the appropriate changes in the next commit.

mjste commented on skfp/similarities/tanimoto.py:
```diff
@@ -0,0 +1,229 @@
+) -> float:
+    """
+    Computes the Tanimoto similarity [1]_ for binary data between two input arrays
+    or sprase matrices. Calls suitable private method based on input type.
```
Typo: Sparse

mjste commented on skfp/similarities/tanimoto.py:
```diff
@@ -0,0 +1,229 @@
+    ----------
+    A : np.darray or scipy.sparse.csr_matrix
+        First binary input array or sparse matrix.
+    B : np.darray or scipy.sparse.csr_matrix
```
Probably ndarray instead of darray

mjste commented on skfp/similarities/tanimoto.py:
```diff
@@ -0,0 +1,229 @@
+        return 1.0
+
+    dot_product_AB: float = np.dot(A, B)
+    dot_product_AA: float = np.sum(A)
```
Not sure, but as far as I remember, dot product of the same vectors is sum of squares. Change it or rename variable please

mjste commented on skfp/similarities/tanimoto.py:
```diff
@@ -0,0 +1,229 @@
+    tanimoto_similarity : float
+        Tanimoto similarity between A and B.
+
+    References
```
Add usage example

mjste commented on skfp/similarities/tanimoto.py:
```diff
@@ -0,0 +1,229 @@
+
+    Parameters
+    ----------
+    A : np.darray or scipy.sparse.csr_matrix
```
Typo

mjste commented on skfp/similarities/tanimoto.py:
```diff
@@ -0,0 +1,229 @@
+    float
+        Tanimoto distance between A and B.
+    """
+
```
Add usage example

j-adamczyk commented on tests/similarities/tanimoto.py:
```diff
@@ -0,0 +1,347 @@
+        (csr_matrix([0, 1, 0, 1]), csr_matrix([0, 1, 1, 0])),
+    ],
+)
+def test_binary_against_sklearn_nn(A, B):
```
`test_sklearn_nearest_neighbors_compatible`

StaniszewskiA commented on tests/similarities/tanimoto.py:
```diff
@@ -0,0 +1,347 @@
+
+    with pytest.raises(
+        TypeError,
+        match=(
+            "Both A and B must be of the same type: either numpy.ndarray or scipy.sparse.csr_matrix, "
+            "got <class 'int'> and <class 'float'>"
+        ),
+    ):
```
That's nice, I didn't now you could do that in Python

StaniszewskiA commented on skfp/similarities/tanimoto.py:
```diff
@@ -0,0 +1,229 @@
+from scipy.sparse import csr_matrix
+from sklearn.utils._param_validation import validate_params
+
+
```
@validate_params does  [https://scikit-learn.org/stable/glossary.html#term-array-like](not count sparse arrays) as **"array-like"**. Do you know a way around it?

j-adamczyk replied:
`["array-like", csr_array]` should work

StaniszewskiA commented on skfp/similarities/tanimoto.py:
```diff
@@ -0,0 +1,229 @@
+    dot_product_BB: float = B.sum()
+
+    denominator: float = dot_product_AA + dot_product_BB - dot_product_AB
+    _check_zero_denominator(denominator)
```
It does not raise any errors. I will add another test for all-ones vectors though.

j-adamczyk replied:
So in what situation this function would raise an error? I mean, the only thing `_check_zero_denominator` does is to raise an error. When two vectors have nothing in common, e.g. `[0, 1, 1]` and `[1, 0, 0]`? In that situation we should return 0.

StaniszewskiA commented:
I've implemented all of the suggested changes, but I still need to figure out a proper docs rendering.

j-adamczyk commented on skfp/similarities/tanimoto.py:
```diff
@@ -0,0 +1,350 @@
+    """
+    Calculates the Tanimoto similarity between two count data numpy arrays.
+    """
+    _check_nan(vec_a)
+    _check_nan(vec_b)
+
+    if np.sum(vec_a) == 0 and np.sum(vec_b) == 0:
+        return 1.0
+
+    dot_ab: float = np.dot(vec_a, vec_b)
+    dot_aa: float = np.dot(vec_a, vec_a)
+    dot_bb: float = np.dot(vec_b, vec_b)
+
+    denominator: float = dot_aa + dot_bb - dot_ab
+    _check_zero_denominator(denominator)
+
+    tanimoto_sim: float = dot_ab / denominator
+
+    return tanimoto_sim
```
It would be beneficial to consider [Numba JIT compilation](https://numba.readthedocs.io/en/stable/user/jit.html) to speed up such functions. Could you measure this, e.g. for nearest neighbors searches on HIV dataset and ECFP4 fingerprints?

j-adamczyk commented on skfp/similarities/tanimoto.py:
```diff
@@ -0,0 +1,350 @@
+    if np.sum(vec_a) == 0 and np.sum(vec_b) == 0:
+        return 1.0
+
+    dot_ab: float = vec_a.multiply(vec_b).sum()
+    dot_aa: float = vec_a.multiply(vec_a).sum()
+    dot_bb: float = vec_b.multiply(vec_b).sum()
+
+    denominator: float = dot_aa + dot_bb - dot_ab
+    _check_zero_denominator(denominator)
+
+    tanimoto_sim: float = dot_ab / denominator
```
I think that if you use `__mul__` instead of `multiply`, you can use the same function for both NumPy and SciPy

StaniszewskiA replied:
We need to keep these methods separate since we want to use JIT with the NumPy one

StaniszewskiA commented:
Since **_tanimoto_count_numpy** is JIT-compilated, tests covering it do not count towards overall coverage; hence, I can't raise it above 85%. Is that okay?

https://github.com/numba/numba/pull/5660

j-adamczyk commented:
Sure, no problem, test coverage is not a goal in itself.

StaniszewskiA commented:
I've added JIT compilation wherever I could and fixed the documentation rendering. Take a look at **tanimoto_binary_similarity**:

1. The SciPy implementation requires using **.toarray().ravel()** on the input array. I think this misses the point of using sparse arrays, so I implemented it myself for this case.
2. The SciPy implementation returns Jaccard Distance, so I needed to subtract the result from one. Alternatively, we could implement this logic within the **tanimoto_binary_distance** function and return 1−distance inside **tanimoto_binary_similarity**. However, I think the current implementation is more consistent with the rest of the code.

j-adamczyk commented on skfp/distances/tanimoto.py:
```diff
@@ -0,0 +1,317 @@
+
+    Computes the Tanimoto distance for binary data between two input arrays or sparse matrices
+    by subtracting the similarity value from 1. The calculated distance falls within
+    the range [0, 1], with passing all-zero vectors to this function resulting in a distance of 0.
```
End sentence after `[0, 1]` (BTW, it should be in backticks, so it renders nicely), and start next sentence as "Passing all-zero..."

j-adamczyk commented on skfp/distances/tanimoto.py:
```diff
@@ -0,0 +1,317 @@
+
+        sim(vec_a, vec_b) = \\frac{vec_a \\cdot vec_b}{\\|vec_a\\|^2 + \\|vec_b\\|^2 - vec_a \\cdot vec_b}
+
+    Calculated similarity falls within the range of 0-1.
+    Passing all-zero vectors to this function result in similarity of 1.
```
Make the description more similar in format to binary variant

j-adamczyk commented on skfp/distances/tanimoto.py:
```diff
@@ -0,0 +1,317 @@
+    vec_a: Union[np.ndarray, csr_array], vec_b: Union[np.ndarray, csr_array]
+) -> float:
+    """
+    Computes the Tanimoto distance for binary data between two input arrays or sparse matrices
+    by subtracting similarity value from 1.
+    Calculated distance falls within the range from 0 to 1.
+    Passing all-zero vectors to this function results in distance of 0.
```
Add formula, unify description format with the ones above

j-adamczyk commented on skfp/distances/tanimoto.py:
```diff
@@ -0,0 +1,317 @@
+
+@njit(parallel=True)
+def _tanimoto_count_numpy(vec_a: np.ndarray, vec_b: np.ndarray) -> float:
+    """
+    Calculates the Tanimoto similarity between two count data numpy arrays.
+    This function uses Numba for JIT compilation and parallel computations to
+    efficiently compute the similarity between two vectors.
+    """
```
I don't think that docstring is necessary here, the function is very readable already

j-adamczyk commented on skfp/distances/tanimoto.py:
```diff
@@ -0,0 +1,317 @@
+
+
+def _tanimoto_count_scipy(vec_a: csr_array, vec_b: csr_array) -> float:
+    """
+    Calculates the Tanimoto similarity between two count data scipy arrays.
+    """
```
Same as above, unnecessary in my opinion

j-adamczyk commented on skfp/distances/tanimoto.py:
```diff
@@ -0,0 +1,317 @@
+    dot_aa: float = vec_a.multiply(vec_a).sum()
+    dot_bb: float = vec_b.multiply(vec_b).sum()
+
+    denominator: float = dot_aa + dot_bb - dot_ab
+
+    tanimoto_sim: float = dot_ab / denominator
+
+    return tanimoto_sim
```
I think you can just return a formula:
```
return dot_ab / (dot_aa + dot_bb - dot_ab)
```

j-adamczyk commented on skfp/distances/__init__.py:
```diff
@@ -0,0 +1,8 @@
+"""Classes for efficient distance and similarity calculations."""
```
Maybe without `efficient`, since everything should be efficient

j-adamczyk commented on skfp/distances/tanimoto.py:
```diff
@@ -0,0 +1,321 @@
+
+
+    The calculated similarity falls within the explicit range `[0, 1]`.
+    Passing all-zero vectors to this function resulting in a similarity of 1.
```
results

j-adamczyk commented on skfp/distances/tanimoto.py:
```diff
@@ -0,0 +1,321 @@
+    .. math::
+
+        sim(vec_a, vec_b) = \\frac{|vec_a \\cap vec_b|}{|vec_a| + |vec_b| - |vec_a \\cap vec_b|}
+
+
```
2 empty lines are required, or would just 1 work?

j-adamczyk commented on skfp/distances/tanimoto.py:
```diff
@@ -0,0 +1,321 @@
+        dist(vec_a, vec_b) = 1 - sim(vec_a, vec_b)
+
+    The calculated distance falls within the range `[0, 1]`.
+    Passing all-zero vectors to this function resulting in a distance of 0.
```
results

j-adamczyk commented on skfp/distances/tanimoto.py:
```diff
@@ -0,0 +1,321 @@
+        sim(vec_a, vec_b) = \\frac{vec_a \\cdot vec_b}{\\|vec_a\\|^2 + \\|vec_b\\|^2 - vec_a \\cdot vec_b}
+
+    Calculated similarity falls within the range of `[0, 1]`.
+    Passing all-zero vectors to this function result in similarity of 1.
```
results

j-adamczyk reviewed (approved):
Just fix docstring tests with imports

mjste commented on skfp/distances/tanimoto.py:
```diff
@@ -0,0 +1,332 @@
+    vec_a: Union[np.ndarray, csr_array], vec_b: Union[np.ndarray, csr_array]
+) -> float:
+    """
+    Tanimoto simlarity for vectors of binary values.
```
Typo

j-adamczyk replied:
+1

mjste commented on skfp/distances/tanimoto.py:
```diff
@@ -0,0 +1,332 @@
+    Parameters
+    ----------
+    vec_a : {ndarray, sparse matrix}
+        First binary input array or sparse matrix.
```
Here we are in body of count tanimoto... Shouldn't it be count for instead of binary input?

j-adamczyk replied:
+1

mjste commented:
What's the speedup caused by introducing numba?

j-adamczyk commented on skfp/distances/tanimoto.py:
```diff
@@ -0,0 +1,332 @@
+    vec_a: Union[np.ndarray, csr_array], vec_b: Union[np.ndarray, csr_array]
+) -> float:
+    """
+    Tanimoto simlarity for vectors of count values.
```
Typo

StaniszewskiA commented:
> What's the speedup caused by introducing numba?


Without Numba:
![tmp_b64af1b4-8de1-4feb-aff5-5b7c1e88d659.png](https://github.com/user-attachments/assets/ef092c41-6149-42c6-b55e-293ffb43aba5)

With Numba:
![tmp_695aa55a-720f-4cbe-b350-ec05289e6173.png](https://github.com/user-attachments/assets/0f206adb-f21c-4376-afd1-78132e61773a)

I forgot to change the labels, so please ignore them. These plots show the running time of two attempts with data from the HIV dataset.

StaniszewskiA commented:
I hope there are no typos remaining.

mjste reviewed (changes requested):
I wanted to check docs rendering and such problem occured:
 
checking consistency... /home/mjste/Desktop/astanfingerprints/docs/modules/similarities_and_distances.rst: WARNING: document isn't included in any toctree

After opening index.html I don't see these modules. Missing entry in docs/modules/classes.rst
